service/contractsv1/api.go

```diff
@@ -0,0 +1,194 @@
+//      The search always ends at 23:59:59 UTC of the specified date. The default end date is the current date.
+func (c *Contractsv1) ListProductsPerContract(contractID, from, to string) (*OutputContracts, error) {
+	query := map[string]string{}
+	if contractID == "" {
+		return nil, fmt.Errorf("Missing argument 'contractID'")
+	}
+
+	apiURI := fmt.Sprintf("%s/contracts/%s/products/summaries", basePath, contractID)
+
+	if from != "" {
+		query["from"] = from
+	}
+
+	if to != "" {
+		query["to"] = to
+	}
+
+	// Create and execute request
+	resp, err := c.Client.Rclient.R().
+		SetResult(OutputContracts{}).
+		SetError(ContractsErrorv1{}).
+		SetQueryParams(query).
+		Get(apiURI)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.IsError() {
+		e := resp.Error().(*ContractsErrorv1)
+
+		return nil, e
+	}
+
+	return resp.Result().(*OutputContracts), nil
+}
+
+// ListReportingGroups gets the IDs of the Content Provider (CP) reporting groups that you have access to along with their names.
+// To run this operation, your user account needs the CPCode Rep Group role.
+func (c *Contractsv1) ListReportingGroups() (*OutputReportingGroups, error) {
+	apiURI := fmt.Sprintf("%s/reportingGroups/", basePath)
+
+	// Create and execute request
+	resp, err := c.Client.Rclient.R().
+		SetResult(OutputReportingGroups{}).
+		SetError(ContractsErrorv1{}).
+		Get(apiURI)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.IsError() {
+		e := resp.Error().(*ContractsErrorv1)
+
+		return nil, e
+	}
+
+	return resp.Result().(*OutputReportingGroups), nil
+}
+
+// ListReportingGroupIDs gets the IDs of the Content Provider (CP) reporting groups that you have access to.
+// To run this operation, your user account needs the CPCode Rep Group role.
+func (c *Contractsv1) ListReportingGroupIDs() (*OutputReportingGroupIDs, error) {
+	apiURI := fmt.Sprintf("%s/reportingGroups/identifiers", basePath)
+
+	// Create and execute request
+	resp, err := c.Client.Rclient.R().
+		SetResult(OutputReportingGroupIDs{}).
+		SetError(ContractsErrorv1{}).
+		Get(apiURI)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.IsError() {
+		e := resp.Error().(*ContractsErrorv1)
+
+		return nil, e
+	}
+
+	return resp.Result().(*OutputReportingGroupIDs), nil
+}
+
+// ListProductsPerReportingGroup gets the IDs and names of the products associated with the reporting group for the time frame selected.
+// To run this operation, your user account needs the CPCode Rep Group role.
+// When a request is successful, it may return either a 200 or a 300 response.
+// The API returns a 200 (OK) response when the CP code reporting group is associated with only one contract.
+// You’ll receive a 300 (Multiple Choices) response when the request returns a list of matching contracts because
+// the CP code reporting group is associated with multiple contracts.
+// To retrieve product information when you receive a 300 response code, make a new GET request to the hyperlinks provided in the response.
```

Comment:
🤔 We need to make request towards response hyperlinks ?

Reply:
Yes, if you will get 300

Reply:
I think it is  something we should not be doing in API. IMHO if  that happens a result of specific type should be returned specific to this particular situation with a flag indicating a different action should be taken by hugher level api

Reply:
This is why I return 2 values of certain types and error. And based on result you can decide what to do in higher level API.
```
products, contracts, err := api.ListProductsPerReportingGroup(ID, "", "")

if contracts != nil {...}
```

Reply:
The root thing here is `make a new GET request to the hyperlinks provided in the response` which I would not to expect in API system.
So based on the above must I do `contracts[0].Hyperlink` and GET ?

Reply:
Ok - makes sense - then just for our awesomness ;) change the description as it implies we implemented API in very strange way